resolwe/test_helpers/test_runner.py

```diff
@@ -5,38 +5,298 @@
+        self.args = args
+        self.kwargs = kwargs
+
+    def __enter__(self):
+        """Enter the ``with`` context."""
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """Exit the context and call the saved callable."""
+        self.call(*self.args, **self.kwargs)
+        return False
+
+
+def _manager_setup():
+    """Execute common setup operations."""
+    # run executor listener first, so there's no startup race condition where a manager fires
+    # off an executor too early for the standalone manager to respond
+    db.connections.close_all()
+    state.update_constants()
+    manager.reset()
+
+
+def _sequence_paths(paths):
+    """Extend the last components of the given paths with a number.
+
+    The method finds the lowest number such that all given paths, when
+    extended by it, are unique and can be created. The paths are then
+    also created.
+
+    :param paths: The list of paths to be extended and created.
+    :return: The list of created paths.
+    """
+    seq = 0
+    while True:
+        # Note for parallel execution: infinite zigzagging ladders are
+        # not possible, because the directories are always created in
+        # the same order. The problem would be if process A succeeded
+        # in creating data/test_1, but process B would beat it to
+        # upload/test_1 (so that both would roll back and continue
+        # with _2, etc.). For B to succeed in creating upload/test_1,
+        # it must have already succeeded in creating data/test_1,
+        # meaning A could not possibly have succeeded with data/test_1.
+        seq += 1
+        created = []
+
+        for base_path in paths:
+            head, tail = os.path.split(base_path)
+            path = os.path.join(head, '{}_{}'.format(tail, seq))
+            try:
+                os.makedirs(path)
+                created.append(path)
+            except OSError:
+                break
+
+        if len(created) == len(paths):
+            return created
+
+        # If they're not equal, we failed and need to roll back;
+        # errors are entirely irrelevant here, removal is purely
+        # best effort.
+        for path in created:
+            try:
+                os.rmdir(path)
+            except:  # pylint: disable=bare-except
+                pass
+
+
+def _create_test_dirs():
+    """Create all the testing directories."""
+    # Due to the unique way the docker mappings are updated, we need
+    # the pre-fudge settings copied and sheltered for the updater to work.
+    previous_settings = copy.copy(resolwe_settings.FLOW_EXECUTOR_SETTINGS)
+    items = ['DATA_DIR', 'UPLOAD_DIR', 'RUNTIME_DIR']
+    paths = _sequence_paths([resolwe_settings.FLOW_EXECUTOR_SETTINGS[i] for i in items])
+    for item, path in zip(items, paths):
+        resolwe_settings.FLOW_EXECUTOR_SETTINGS[item] = path
+    resolwe_settings._update_flow_docker_mappings(previous_settings)  # pylint: disable=protected-access
```

Comment:
It does update the settings, and you've fallen precisely into the trap I mentioned during the meeting. Renaming the function to `_get_updated_*` would lead people to believe that this function is a getter, which it isn't. Changing the code to your suggestion would lead to unwanted interplay with the way Python name binding works, leading to general breakage and, most importantly, to _not_ updating the docker mappings.